src/dbnode/storage/index/block.go

```diff
@@ -884,6 +919,190 @@ func (b *block) addQueryResults(
 	return batch, size, err
 }
 
+// Aggregate acquires a read lock on the block so that the segments
+// are guaranteed to not be freed/released while accumulating results.
+// This allows references to the mmap'd segment data to be accumulated
+// and then copied into the results before this method returns (it is not
+// safe to return docs directly from the segments from this method, the
+// results datastructure is used to copy it every time documents are added
+// to the results datastructure). This is similar to how Query() operates.
+// NB: Aggregate is required in addition to Query, to optimise for the case
+// when we can skip going to raw documents, and instead rely on pre-aggregated
+// results via the FST underlying the index.
+func (b *block) Aggregate(
+	cancellable *resource.CancellableLifetime,
+	opts QueryOptions,
+	results AggregateResults,
+) (bool, error) {
+	b.RLock()
+	defer b.RUnlock()
+
+	if b.state == blockStateClosed {
+		return false, ErrUnableToQueryBlockClosed
+	}
+
+	segs, err := b.segmentsWithRLock()
+	if err != nil {
+		return false, err
+	}
+
+	size := results.Size()
+	batch := b.aggEntryPool.Get()
+	batchSize := cap(batch)
+	if batchSize == 0 {
+		batchSize = defaultAggregateResultsEntryBatchSize
+	}
+
+	aggOpts := results.AggregateResultsOptions()
+	iterateTerms := aggOpts.Type == AggregateTagNamesAndValues
+	defer func() {
+		b.aggEntryPool.Put(batch)
+	}()
+
+	iter, err := newFieldsAndTermsIterator(nil, fieldsAndTermsIteratorOpts{})
+	if err != nil {
+		return false, err
+	}
+
+	for _, s := range segs {
+		if opts.LimitExceeded(size) {
+			break
+		}
+
+		err = iter.Reset(s, fieldsAndTermsIteratorOpts{
+			iterateTerms: iterateTerms,
+			allowFn: func(field []byte) bool {
+				// skip any field names that we shouldn't allow.
+				if bytes.Equal(field, doc.IDReservedFieldName) {
+					return false
+				}
+				return aggOpts.TermFilter.Allow(field)
+			},
+		})
+		if err != nil {
+			return false, err
+		}
+
+		for iter.Next() {
+			if opts.LimitExceeded(size) {
+				break
+			}
+
+			field, term := iter.Current()
+			batch = b.appendAggregateResults(batch, field, term, iterateTerms)
+			if len(batch) < batchSize {
+				continue
+			}
+
+			batch, size, err = b.addAggregateResults(cancellable, results, batch)
+			if err != nil {
+				iter.Close()
+				return false, err
+			}
+		}
+
+		if err := iter.Err(); err != nil {
+			iter.Close()
+			return false, err
+		}
+
+		if err := iter.Close(); err != nil {
```

Comment:
nit: I don't think this is necessary; you reset every loop and close the iterator explicitly below already

Reply:
Hm it's necessary cause we want to ensure we're releasing any internal state (via Close()). Reset() just clears that state, doesn't release it.

Reply:
Fair; in that case it's a little weird that we double close below

For this particular iterator it's not really a problem, but it's plausible for an iter implementatiopn to panic if it's double closed here

Reply:
Definitely a fair point. This implementation doesn't have that issue intentionally for this reason - so as to ensure we free resources. I'll make a note about this property in a comment.

Reply:
done

Reply:
Still feels a little sketch to me because I think some of our `Close()` methods return objects to pools

Reply:
that's a fair point. I'll rework to not rely on this behaviour.

Reply:
done